Improve assertion message in `AlignmentDirectional.resolve`

SalehTZ opened the pull request:
Replaces the generic assertion in `AlignmentDirectional.resolve` and `_MixedAlignment.resolve` with
`debugCheckCanResolveTextDirection`.

This change provides a more helpful and descriptive error message when the `TextDirection` is null, improving the developer experience by clearly stating why the resolution failed and how to fix it.

Fixes: #171813, Fixes: #171814, also fixes: #171812

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

SalehTZ committed:
feat: Improve assertion message in AlignmentDirectional.resolve

Replaces the generic assertion in `AlignmentDirectional.resolve` and `_MixedAlignment.resolve` with
`debugCheckCanResolveTextDirection`.

This change provides a more helpful and descriptive error message when the `TextDirection` is null, improving the developer
experience by clearly stating why the resolution failed and how to fix it.

Fixes: #171813

SalehTZ committed:
test: Verify assertion for null TextDirection in AlignmentDirectional

Adds a test case to ensure that calling `AlignmentDirectional.resolve` with a null `TextDirection` throws a `FlutterError`
with the expected improved message.

SalehTZ committed:
Merge branch 'master' into add-detailed-error-message-for-AlignmentDirectional.resolve

SalehTZ committed:
test: Verify assertion for null TextDirection in `_MixedAlignment.resolve`

Adds a test case that uses `add()` method to ensure that calling `_MixedAlignment.resolve` with a null `TextDirection` throws a `FlutterError`
with the expected improved message.

SalehTZ committed:
fix: removed method which was added by mistake

nt4f04uNd commented on packages/flutter/lib/src/painting/alignment.dart:
```diff
@@ -584,7 +585,7 @@ class AlignmentDirectional extends AlignmentGeometry {
   @override
   Alignment resolve(TextDirection? direction) {
-    assert(direction != null, 'Cannot resolve $runtimeType without a TextDirection.');
+    assert(debugCheckCanResolveTextDirection(direction, '$runtimeType'));
```
It's better to use `'$AlignmentDirectional'` here, because `runtimeType` is a slow operation

It's ok to use it inside the assertion function itself [like here](https://github.com/flutter/flutter/blob/a04fb324be734cf18811c30c06baf9bf07b3bab3/packages/flutter/lib/src/rendering/flex.dart#L1337) - that way it will be stripped from release builds, but in our case the call to it will happen even with assertions disabled.

SalehTZ replied:
done.

nt4f04uNd commented on packages/flutter/lib/src/painting/alignment.dart:
```diff
@@ -653,7 +654,7 @@ class _MixedAlignment extends AlignmentGeometry {
   @override
   Alignment resolve(TextDirection? direction) {
-    assert(direction != null, 'Cannot resolve $runtimeType without a TextDirection.');
+    assert(debugCheckCanResolveTextDirection(direction, '$runtimeType'));
```
Same here

SalehTZ replied:
done.

SalehTZ committed:
replaced `$runtimeType` with`$AlignmentDirectional`

more details: https://github.com/flutter/flutter/pull/172096#discussion_r2217762492

SalehTZ committed:
fix: incorrect variable was used

victorsanni reviewed (approved):
LGTM, thanks for the PR.

MitchellGoodwin reviewed (approved):
LGTM as well!
